libtsuba/include/tsuba/Cache.h

```diff
@@ -0,0 +1,182 @@
+
+template <typename Value>
+class KATANA_EXPORT Cache {
+public:
+  using ListType = std::list<KeyBase>;
+  struct MapValue {
+    Value value;
+    // This allows us to delete the old position in the LRU list without a scan
+    typename ListType::iterator lruit;
+  };
+  using MapType = std::unordered_map<KeyBase, MapValue, KeyBase::Hash>;
+  enum class ReplacementPolicy { kLRUSize, kLRUBytes };
+
+  Cache(
+      ReplacementPolicy policy = ReplacementPolicy::kLRUSize,
+      size_t lru_capacity = 0, size_t bytes_capacity = 0,
+      std::function<void(const KeyBase& key)> evict_cb = nullptr,
+      std::function<size_t(const Value& value)> value_to_bytes = nullptr)
+      : policy_(policy),
+        lru_capacity_(lru_capacity),
+        bytes_capacity_(bytes_capacity),
+        evict_cb_(std::move(evict_cb)),
+        value_to_bytes_(std::move(value_to_bytes)) {
+    KATANA_LOG_VASSERT(
+        policy_ != ReplacementPolicy::kLRUSize || lru_capacity_ > 0,
+        "kLRUSize policy requires non-zero LRU entries");
+    KATANA_LOG_VASSERT(
+        policy_ != ReplacementPolicy::kLRUBytes || bytes_capacity_ > 0,
+        "kLRUBytes policy requires non-zero byte capacity");
+  }
+
+  size_t size() const { return key_to_value_.size(); }
+  size_t bytes() const { return total_bytes_; }
+
+  size_t LRUSizeCapacity() const { return lru_capacity_; }
+  size_t LRUBytesCapacity() const { return bytes_capacity_; }
+
+  bool Empty() const { return key_to_value_.empty(); }
+
+  bool Contains(const KeyBase& key) {
+    return key_to_value_.find(key) != key_to_value_.end();
+  }
+
+  void Insert(const KeyBase& key, const Value& value) {
+    auto mapit = key_to_value_.find(key);
+    if (mapit == key_to_value_.end()) {
+      lru_list_.push_front(key);
+      key_to_value_[key] = {value, lru_list_.begin()};
+      if (value_to_bytes_ != nullptr) {
+        total_bytes_ += value_to_bytes_(value);
+      }
+    } else {
+      mapit->second.value = value;
+      update_lru(mapit);
+    }
+    evict_if_necessary();
+  }
+
+  std::optional<Value> Get(const KeyBase& key) {
+    // lookup value in the cache
+    std::optional<Value> ret = std::nullopt;
+    auto it = key_to_value_.find(key);
+    if (it != key_to_value_.end()) {
+      ret = update_lru(it);
+    }
+    return ret;
+  }
+
+private:
+  Value update_lru(typename MapType::iterator mapit) {
+    auto lruit = mapit->second.lruit;
+    auto lru_head = lru_list_.begin();
+    if (lruit != lru_head) {
+      // move item to the front of the most recently used list
+      lru_list_.splice(lru_head, lru_list_, lruit);
+      mapit->second.lruit = lru_head;
+    }
+    return mapit->second.value;
+  }
+
+  void evict_one() {
+    KeyBase evicted_key;
+    // evict item from the end of most recently used list
+    auto tail = --lru_list_.end();
+    KATANA_LOG_ASSERT(tail != lru_list_.end());
+    evicted_key = *tail;
+    lru_list_.erase(tail);
+    auto evicted_value = key_to_value_.at(evicted_key).value;
```

Comment:
Consider some moves here to avoid copying. Though overall, the cache interface seems to assume copying is cheap. Is this intentional?

Reply:
I added some std::move's because I'm bad at knowing when those will be important.

I have this comment now at the top of the file.

```
// This is single threaded only, it is not intended to store large objects,
// but rather metadata (e.g., a shared_ptr to a property column).
```